### PR TITLE
Incorporate feedback for the AI@2025

### PR DESCRIPTION
**Pull request by edyoshikun**
This PR addresses #10 with following : 

- [x] Remove Part 2 ( GAN approach for virtual staining). Note: we had already removed this in #9 . 
- [x] Update the `README.md` to remove the `Part 1` and `Part2` since we are consolidating this exercise.
- [x] Test and update the dependencies
- [x] Mirror the demo we have in the VCP here. 
- [x] Opened a paired PR in [VisCy 128](https://github.com/mehta-lab/VisCy/pull/278). 
- [x]  Address Ziwen, Dian, and Albert's review

### CODE REVIEW

File: solution.py [hunk 1]
```diff
@@ -927,24 +937,24 @@ def log_batch_jupyter(batch):
 # %% [markdown] tags=[]
```
Comment (AlbertDominguez) [inline solution.py:930]:
```suggestion
# %% [markdown] tags=[]
<div class="alert alert-danger">
If at any point during the exercise you need to reload your trained model, simply run the following code:

```python
# phase2fluor_model_ckpt = natsorted(glob(
#  f"{os.getcwd()}/data/06_image_translation/logs/phase2fluor/version_0/checkpoints/*.ckpt")
# )[-1]
# phase2fluor_config = dict(
#     in_channels=1,
#     out_channels=2,
#     encoder_blocks=[3, 3, 9, 3],
#     dims=[96, 192, 384, 768],
#     decoder_conv_blocks=2,
#     stem_kernel_size=(1, 2, 2),
#     in_stack_depth=1,
#     pretraining=False,
# )
# phase2fluor_model = VSUNet.load_from_checkpoint(
#     phase2fluor_model_ckpt,
#     architecture="UNeXt2_2D",
#     model_config = phase2fluor_config,
#     accelerator='gpu'
# )
# ```
# </div>

# %% [markdown] tags=[]
```

File: solution.py [hunk 2]
```diff
@@ -1365,8 +1589,11 @@ def min_max_scale(image:ArrayLike)->ArrayLike:
 
         ###### Segmentation based metrics #########
         # Load the manually curated nuclei target label
-        print('Computing Segmentation Metrics')
-        pred_label,target_label= cellpose_segmentation(predicted_nuc_phase2fluor,target_nucleus)
+        pbar.set_description(f"Processing FOV {fov} - Computing Segmentation Metrics")
+        pbar.refresh()
+        pred_label, target_label = cellpose_segmentation(
+            predicted_nuc_phase2fluor, target_nucleus
```
Comment (AlbertDominguez) [inline solution.py]:
```suggestion
            predicted_nuc_phase2fluor, target_nucleus[0]
```
`target_nucleus` is 4-dimensional here (`(1,1,2048,2048)`), which tricks Cellpose into thinking this is a 3D stack and an error is raised.

File: solution.py [hunk 3]
```diff
@@ -1378,19 +1605,31 @@ def min_max_scale(image:ArrayLike)->ArrayLike:
...
+            "mAP": coco_metrics["map"].item(),
+            "mAP_50": coco_metrics["map_50"].item(),
+            "mAP_75": coco_metrics["map_75"].item(),
+            "mAR_100": coco_metrics["mar_100"].item(),
         }
 
-        pred_label,target_label= cellpose_segmentation(predicted_nuc_pretrained,target_nucleus)
-        
+        pred_label, target_label = cellpose_segmentation(
+            predicted_nuc_pretrained, target_nucleus
```
Comment (AlbertDominguez) [inline solution.py]:
```suggestion
            predicted_nuc_pretrained, target_nucleus[0]
```
`target_nucleus` is 4-dimensional here (`(1,1,2048,2048)`), which tricks Cellpose into thinking this is a 3D stack and an error is raised.

**Review by AlbertDominguez — CHANGES_REQUESTED** [review]
Hey @edyoshikun,

It's looking great, and seems to mostly run on the VMs! The main issue I found is that the cell computing SSIM/Dice seems to break for me, due to an unexpected dim given to cellpose. I've suggested the change.

I also had to play around a bit with the paths to the downloaded data. Are they all gonna download the data, or they'll use it from the shared mount? 

Here's my other minor feedback/suggestions:

- In the `log_batch_jupyter`, it'd be nice to add `cmap=gray` to display the images for coherency with the tensorboard viz.
- In the alert containing several code snippets, It'd be nice to add the line to load phase2fluor model user trains in case there's a crash, laptop sleeps, nb disconnection, etc. (I made the suggestion, feel free to accept/ignore!)

Lmk if anything is unclear 🙂

File: solution.py (hunk 1 above)
Comment (edyoshikun) [reply solution.py:930]: thanks. added this extra

File: solution.py
Comment (adjavon) [inline solution.py:1]: A couple of quick notes:
- The import cell took surprisingly long to run (~45s)
- To use `iohub` in the terminal (in VSCode) they will also have to set the environment there. We could have them select the interpreter (in VSCode preferences)?
- In "Visualize segmentation comparison: Fluorescence vs Virtual Staining vs Pretrained" it is getting: "Resizing is depricated in v4.0.1+" a few times --> this seems to be a cellpose issue?

File: solution.py [hunk 4]
```diff
@@ -492,22 +503,22 @@ def log_batch_jupyter(batch):
...
 # </div>
 # %% tags=["task"]
 # Here we turn on data augmentation and rerun setup
 # #######################
 # ##### TODO ########
 # #######################
-# HINT: Run dataset.channel_names 
+# HINT: Run dataset.channel_names
 source_channel = ["TODO"]
 target_channel = ["TODO", "TODO"]
```
Comment (adjavon) [inline solution.py]:
```suggestion
```

File: solution.py [hunk 5]
```diff
@@ -492,22 +503,22 @@ def log_batch_jupyter(batch):
...
 # `viscy.transforms` [here](https://github.com/mehta-lab/VisCy/blob/main/viscy/transforms.py). You can look at the docs by running `RandAffined?`.<br><br>
-# *Note these are MONAI transforms that have been redefined for VisCy.* 
+# *Note these are MONAI transforms that have been redefined for VisCy.*
 # [Compare your choice of augmentations by dowloading the pretrained models and config files](https://github.com/mehta-lab/VisCy/releases/download/v0.1.0/VisCy-0.1.0-VS-models.zip).
 # </div>
 # %% tags=["task"]
 # Here we turn on data augmentation and rerun setup
 # #######################
 # ##### TODO ########
 # #######################
```
Comment (adjavon) [inline solution.py]:
```suggestion
```

File: solution.py [hunk 6]
```diff
@@ -662,7 +670,7 @@ def log_batch_jupyter(batch):
 # - Run the next cells to instantiate data module and trainer.
 #   - Add the source channel name and the target channel names
 # - Start the training <br>
-# 
+#
 # <b> Note </b> <br>
 # See ``viscy.unet.networks.Unet2D.Unet2d`` ([source code](https://github.com/mehta-lab/VisCy/blob/7c5e4c1d68e70163cf514d22c475da8ea7dc3a88/viscy/unet/networks/Unet2D.py#L7)) to learn more about the configuration.
```
Comment (adjavon) [inline solution.py]: This source code doesn't quite match the configuration dictionary; the `in_stack_depth` isn't explained. Should it be linking to the VSUNet instead?

File: setup_student.sh
Comment (AlbertDominguez) [inline setup_student.sh:1]: Note that the setup breaks if the user has an existing env containing `06_image_translation` in its name (see [this issue](https://github.com/ai-mbl/intro_dl/issues/7)). Do you want me to directly push a fix @edyoshikun?

Comment (edyoshikun) [reply setup_student.sh:1]: awesome! thanks. Addressed in the last commit.

File: solution.py (hunk 4 above)
Comment (adjavon) [reply solution.py]: Feel free to ignore, thought maybe we don't want to have them re-defining this multiple times.

**Review by adjavon — COMMENTED** [review]
It seems good to me aside from a few minor comments! There are some types/organization things that I've been adding as I go along. I'll push those once I've merged them with your other changes :)

File: solution.py [hunk 7]
```diff
@@ -1556,24 +1835,524 @@ def min_max_scale(image:ArrayLike)->ArrayLike:
...
+# HINT: Print the keys of the `sample` dictionary
+# HINT: Input should be nuclei and membrane channels concatenated
+# HINT: Target should be the original phase image
+
+fluor_input = ...  # TODO: Source
+target_phase = ...  # TODO: Target
+
+# TODO: Make prediction with the fluorescence to phase model
+with torch.no_grad():
+    predicted_phase = ...
```
Comment (adjavon) [inline solution.py:1942]:
```suggestion
# NOTE: The `fluor2phase_model`, returns a tuple. Select the first item with `[0]`
    predicted_phase = ... 
```

File: solution.py [hunk 8]
```diff
@@ -924,27 +929,55 @@ def log_batch_jupyter(batch):
...
-#))[-1]
-#```
-
-#```python
-#phase2fluor_config = dict(
+#
+# ```python
+# phase2fluor_model_ckpt = natsorted(glob(
+#  str("/mnt/efs/aimbl_2025/data/04_image_translation/pretrained_models/AIMBL_Demo/backup.ckpt")
+# ))[-1]
```
Comment (AlbertDominguez) [inline solution.py]:
```suggestion
# phase2fluor_model_ckpt = "/mnt/efs/aimbl_2025/data/04_image_translation/pretrained_models/AIMBL_Demo/backup.ckpt"
```

File: solution.py [hunk 9]
```diff
@@ -1142,22 +1199,27 @@ def process_image(image):
...
 
-# NOTE: if their model didn't go past epoch 5, lost their checkpoint, or didnt train anything. 
+# NOTE: if their model didn't go past epoch 5, lost their checkpoint, or didnt train anything.
 # Uncomment the next lines
-#phase2fluor_model_ckpt = natsorted(glob(
-#  str(top_dir/"06_image_translation/backup/phase2fluor/version_0/checkpoints/*.ckpt")
-#))[-1]
+# phase2fluor_model_ckpt = natsorted(glob(
+#  str("/mnt/efs/aimbl_2025/data/04_image_translation/pretrained_models/AIMBL_Demo/backup.ckpt")
+# ))[-1]
```
Comment (AlbertDominguez) [inline solution.py]:
```suggestion
# phase2fluor_model_ckpt = "/mnt/efs/aimbl_2025/data/04_image_translation/pretrained_models/AIMBL_Demo/backup.ckpt"
```

File: solution.py [hunk 10]
```diff
@@ -1556,24 +1835,524 @@ def min_max_scale(image:ArrayLike)->ArrayLike:
...
+target_mem = target_tensor[0,1].cpu().numpy()
+
+# Saving the single prediction without TTA for later comparison
+with torch.no_grad():
+    single_pred = phase2fluor_model(source_tensor)
+    single_pred_nuc = single_pred[0, 0].cpu().numpy()
+    single_pred_mem = single_pred[0, 1].cpu().numpy()
+
+# TODO: Define TTA transforms using MONAI as a list of tuples (forward, inverse)
+###### YOUR CODE HERE ######
```
Comment (adjavon) [inline solution.py:2135]: I would give a bigger hint here. Maybe add an example of a forward and inverse transform in the blue box above.

File: solution.py [hunk 11]
```diff
@@ -1556,24 +1835,524 @@ def min_max_scale(image:ArrayLike)->ArrayLike:
...
+
+# %% [markdown] tags=[]
+# <div class="alert alert-info">
+# <h3>Bonus: Test Time Augmentation (TTA)</h3>
 # 
+# Test Time Augmentation is a technique where you apply multiple augmentations to a single test image,
+# make predictions on each augmented version, and then combine the results (usually by averaging).
+#
+# **In this section we will:**
+# <ul>
```
Comment (adjavon) [inline solution.py:2101]:
```suggestion
# <ul>
# <li> Return to the `phase2fluor` model </li>
```

File: solution.py [hunk 12]
```diff
@@ -1556,24 +1835,524 @@ def min_max_scale(image:ArrayLike)->ArrayLike:
...
+
+# TODO: Extract nucleus and membrane predictions
+###### YOUR CODE HERE ######
+tta_pred_nuc = ...
+tta_pred_mem = ...
+
+#%% tags=["task"]
+# TODO: Compare TTA results with single prediction
+# Calculate metrics (SSIM, Pearson correlation) for both approaches. Do not forget to normalize the data range to 0-1.
+###### YOUR CODE HERE ######
```
Comment (adjavon) [inline solution.py]:
```suggestion
###### YOUR CODE HERE ######

# TODO Normalize data range to 0-1
# TODO Calculate metrics
# TODO # TTA prediction metrics

# Print comparison
print("\nMetrics Comparison:")
print(f"{'Metric':<20} {'Single':<10} {'TTA':<10} {'Improvement':<12}")
print("-" * 55)
print(f"{'SSIM Nucleus':<20} {ssim_nuc_single:.3f}     {ssim_nuc_tta:.3f}     {ssim_nuc_tta-ssim_nuc_single:+.3f}")
print(f"{'SSIM Membrane':<20} {ssim_mem_single:.3f}     {ssim_mem_tta:.3f}     {ssim_mem_tta-ssim_mem_single:+.3f}")
print(f"{'Pearson Nucleus':<20} {pearson_nuc_single:.3f}     {pearson_nuc_tta:.3f}     {pearson_nuc_tta-pearson_nuc_single:+.3f}")
print(f"{'Pearson Membrane':<20} {pearson_mem_single:.3f}     {pearson_mem_tta:.3f}     {pearson_mem_tta-pearson_mem_single:+.3f}")
```

Comment (adjavon) [reply solution.py]: Adds a bit of structure :)
